stl/inc/algorithm

```diff
@@ -5035,6 +5046,238 @@ _FwdIt shift_right(_ExPo&&, _FwdIt _First, _FwdIt _Last, _Iter_diff_t<_FwdIt> _P
+            } else {
+                _UResult = _Shift_left_impl(_Get_unwrapped(_First), _Get_unwrapped(_Last), _Pos_to_shift);
+            }
+            _Seek_wrapped(_Result, _UResult);
+            return {_STD move(_First), _STD move(_Result)};
+        }
+
+        template <forward_range _Rng>
+            requires permutable<iterator_t<_Rng>>
+        constexpr borrowed_subrange_t<_Rng> operator()(_Rng&& _Range, range_difference_t<_Rng> _Pos_to_shift) const {
+            _STL_ASSERT(_Pos_to_shift >= 0, "shift count must be non-negative (N4910 [alg.shift]/1)");
+
+            if (_Pos_to_shift == 0) {
+                auto _Last = _Rewrap_iterator(_Range, _Get_final_iterator_unwrapped(_Range));
+                return {_RANGES begin(_Range), _STD move(_Last)};
+            }
+
+            if constexpr (sized_range<_Rng>) {
+                auto _First = _RANGES begin(_Range);
+                if (_RANGES distance(_Range) <= _Pos_to_shift) {
+                    return {_First, _First};
+                }
+
+                auto _UFirst   = _Get_unwrapped(_First);
+                auto _Start_at = _RANGES next(_UFirst, _Pos_to_shift);
+                auto _Result   = _RANGES _Move_unchecked(_STD move(_Start_at), _Uend(_Range), _UFirst).out;
+                return {_STD move(_First), _Rewrap_iterator(_Range, _STD move(_Result))};
+            } else {
+                auto _Result = _Shift_left_impl(_Ubegin(_Range), _Uend(_Range), _Pos_to_shift);
+                return {_RANGES begin(_Range), _Rewrap_iterator(_Range, _STD move(_Result))};
+            }
+        }
+
+    private:
+        template <class _It, class _Se>
+        _NODISCARD static constexpr _It _Shift_left_impl(_It _First, _Se _Last, iter_difference_t<_It> _Pos_to_shift) {
+            _STL_INTERNAL_STATIC_ASSERT(permutable<_It>);
+            _STL_INTERNAL_STATIC_ASSERT(sentinel_for<_Se, _It>);
+            _STL_INTERNAL_CHECK(_Pos_to_shift > 0);
+
+            auto _Start_at = _First;
+            if (_RANGES advance(_Start_at, _Pos_to_shift, _Last) != 0) {
+                return _First;
+            }
+
+            return _RANGES _Move_unchecked(_STD move(_Start_at), _STD move(_Last), _STD move(_First)).out;
+        }
+    };
+
+    inline constexpr _Shift_left_fn shift_left{_Not_quite_object::_Construct_tag{}};
+
+    class _Shift_right_fn : private _Not_quite_object {
+    public:
+        using _Not_quite_object::_Not_quite_object;
+
+        template <permutable _It, sentinel_for<_It> _Se>
+        constexpr subrange<_It> operator()(_It _First, const _Se _Last, iter_difference_t<_It> _Pos_to_shift) const {
+            _STL_ASSERT(_Pos_to_shift >= 0, "shift count must be non-negative (N4910 [alg.shift]/5)");
+
+            _Adl_verify_range(_First, _Last);
+            auto _UFirst = _Get_unwrapped(_First);
+            auto _ULast  = _Get_unwrapped(_Last);
+            auto _Size   = _Size_helper(_UFirst, _ULast);

```

Comment:
I guess for debug performance it would be nice to early return here in the `_Pos_to_shift == 0` case, rather than calling `_Size_helper` etc

Reply:
That's not a pure win, though - it costs an extra branch. (Perhaps optimized away in release mode, but definitely not in debug mode.)

In general, when early-return cases aren't required for correctness, I think they're best avoided by default, unless/until we have evidence that such cases are especially common.

Reply:
It is a pure win, the first thing we do inside the helper function is to check against 0.

This would just hoist that check before the `_Size` calculation